[CSS Math Functions] Correctly serialize abs/sign/exp/log/pow/sqrt root nodes

nt1m opened the pull request:
#### ea22468da561b39617b387235d52c2cc0c012dd1
<pre>
[CSS Math Functions] Correctly serialize abs/sign/exp/log/pow/sqrt root nodes
<a href="https://bugs.webkit.org/show_bug.cgi?id=259012">https://bugs.webkit.org/show_bug.cgi?id=259012</a>
rdar://111948465

Reviewed by Darin Adler.

From the spec <a href="https://drafts.csswg.org/css-values-4/#calc-simplification">https://drafts.csswg.org/css-values-4/#calc-simplification</a>:

&gt; If root is an operator node that’s not one of the calc-operator nodes, and all of its calculation children are numeric values with enough information to compute the operation root represents, return the result of running root’s operation using its children, expressed in the result’s canonical unit.

We now always try to resolve the top-level abs/sign/exp/log/pow/sqrt functions, e.g. abs(-1) now becomes calc(1) instead of staying abs(-1).
This is consistent with calc(abs(-1)) being serialized as calc(1).

* LayoutTests/imported/w3c/web-platform-tests/css/css-values/exp-log-serialize-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/exp-log-serialize.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/hypot-pow-sqrt-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/hypot-pow-sqrt-invalid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/hypot-pow-sqrt-serialize-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/hypot-pow-sqrt-serialize.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/signs-abs-serialize-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/signs-abs-serialize.html:

Re-import WPT with fixes from upstream.

* Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
(WebCore::CSSCalcOperationNode::simplifyNode):
* Source/WebCore/css/calc/CSSCalcOperationNode.h:

Canonical link: <a href="https://commits.webkit.org/265879@main">https://commits.webkit.org/265879@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/54d0250b945ddea78e27ac6650bda2a2e959e9b7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12129 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12476 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12778 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13875 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11704 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12160 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14891 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12490 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14392 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12293 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13099 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10268 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14291 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10385 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11009 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18127 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11466 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11168 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14344 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11659 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9608 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10875 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/10884 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2975 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15201 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11512 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/calc/CSSCalcOperationNode.h:
```diff
@@ -145,7 +145,7 @@ class CSSCalcOperationNode final : public CSSCalcExpressionNode {
 
     void makeTopLevelCalc();
-    bool shouldNotPreserveFunction() const { return isMinOrMaxNode() || isHypotNode(); }
+    bool shouldNotPreserveFunction() const { return isSignNode() || isMinOrMaxNode() || isExpNode() || isHypotNode() || isPowOrSqrtNode(); }
```
Could we define `shouldNotPreserveFunction` in terms of `shouldPreserveFunction` or vice versa? Seems a shame to express the same list twice.

nt1m replied:
We probably could, but `shouldPreserveFunction` will get removed anyway as I fix the serialization for the remaining functions since it's actually undesirable behavior.

I'd stay on the side of caution here and keep it that way for now (and have the individual changes for the other functions slowly remove `shouldPreserveFunction`).

If we really wanted, it would probably look something like:
```
shouldPreserveFunction() { return !shouldNotPreserveFunction() && !isCalcSumNode() && !isCalcProductNode() && m_operator != CalcOperator::Subtract && m_operator != CalcOperator::Divide && m_operator != CalcOperator::Multiply;
```

darinadler replied:
We can wait and look at it after you are done with the individual fixes.

nt1m committed:
[CSS Math Functions] Correctly serialize abs/sign/exp/log/pow/sqrt root nodes
https://bugs.webkit.org/show_bug.cgi?id=259012
rdar://111948465

Reviewed by Darin Adler.

From the spec https://drafts.csswg.org/css-values-4/#calc-simplification:

> If root is an operator node that’s not one of the calc-operator nodes, and all of its calculation children are numeric values with enough information to compute the operation root represents, return the result of running root’s operation using its children, expressed in the result’s canonical unit.

We now always try to resolve the top-level abs/sign/exp/log/pow/sqrt functions, e.g. abs(-1) now becomes calc(1) instead of staying abs(-1).
This is consistent with calc(abs(-1)) being serialized as calc(1).

* LayoutTests/imported/w3c/web-platform-tests/css/css-values/exp-log-serialize-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/exp-log-serialize.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/hypot-pow-sqrt-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/hypot-pow-sqrt-invalid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/hypot-pow-sqrt-serialize-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/hypot-pow-sqrt-serialize.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/signs-abs-serialize-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/signs-abs-serialize.html:

Re-import WPT with fixes from upstream.

* Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
(WebCore::CSSCalcOperationNode::simplifyNode):
* Source/WebCore/css/calc/CSSCalcOperationNode.h:

Canonical link: https://commits.webkit.org/265879@main
